File: software/examples/go/example_callback.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
	"github.com/Tinkerforge/go-api-bindings/servo_v2_bricklet"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Servo Bricklet 2.0.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	s, _ := servo_v2_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	s.RegisterPositionReachedCallback(func(servoChannel uint16, position int16) {
		if position == 9000 {
			fmt.Println("Position: 90°, going to -90°")
			s.SetPosition(servoChannel, -9000)
		} else if position == -9000 {
			fmt.Println("Position: -90°, going to 90°")
			s.SetPosition(servoChannel, 9000)
		} else {
			// Can only happen if another program sets position
			fmt.Println("Error")
		}
	})

	// Enable position reached callback
	s.SetPositionReachedCallbackConfiguration(0, true)

	// Set velocity to 100°/s. This has to be smaller or equal to the
	// maximum velocity of the servo you are using, otherwise the position
	// reached callback will be called too early
	s.SetMotionConfiguration(0, 10000, 500000, 500000)
	s.SetPosition(0, 9000)
	s.SetEnable(0, true)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

	s.SetEnable(0, false)
}

File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{ip_connection::IpConnection, servo_v2_bricklet::*};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Servo Bricklet 2.0.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let s = ServoV2Bricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    let position_reached_receiver = s.get_position_reached_callback_receiver();

    // Spawn thread to handle received callback messages.
    // This thread ends when the `s` object
    // is dropped, so there is no need for manual cleanup.
    let s_copy = s.clone(); // Device objects don't implement Sync, so they can't be shared 
                            // between threads (by reference). So clone the device and move the copy.
    thread::spawn(move || {
        for position_reached in position_reached_receiver {
            if position_reached.position == 9000 {
                println!("Position: 90°, going to -90°");
                s_copy.set_position(position_reached.servo_channel, -9000).recv()?;
            } else if position_reached.position == -9000 {
                println!("Position: -90°, going to 90°");
                s_copy.set_position(position_reached.servo_channel, 9000).recv()?;
            } else {
                // Can only happen if another program sets position
                println!("Error");
            }
        }
    });

    // Enable position reached callback
    s.set_position_reached_callback_configuration(0, true).recv()?;

    // Set velocity to 100°/s. This has to be smaller or equal to the
    // maximum velocity of the servo you are using, otherwise the position
    // reached callback will be called too early
    s.set_motion_configuration(0, 10000, 500000, 500000).recv()?;
    s.set_position(0, 9000).recv()?;
    s.set_enable(0, true).recv()?;

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;

    s.set_enable(0, false).recv()?;

    ipcon.disconnect();
    Ok(())
}
